lxd/ovn.go

```diff
@@ -0,0 +1,157 @@
+}
+
+// NewOVNLogMonitor returns a new OVNLogMonitor.
+func NewOVNLogMonitor(s *state.State) *OVNLogMonitor {
+	// Use the logs from MicroOVN if available, otherwise use the ones in /var/log/ovn.
+	ovnLogPath := "/var/snap/microovn/logs"
+
+	if !shared.PathExists(ovnLogPath) {
+		ovnLogPath = "/var/log/ovn"
+	}
+
+	return &OVNLogMonitor{
+		state: s,
+		paths: []string{
+			filepath.Join(ovnLogPath, "ovn-controller.log"),
+			filepath.Join(ovnLogPath, "ovn-northd.log"),
+			filepath.Join(ovnLogPath, "ovsdb-server-nb.log"),
+			filepath.Join(ovnLogPath, "ovsdb-server-sb.log"),
+		},
+		initialStats:  make(map[string]fs.FileInfo),
+		lastTimestamp: time.Now(),
+	}
+}
+
+// Start starts the log file monitor. It checks whether some file stats have changed, and sends the log entries to the events API.
+func (l *OVNLogMonitor) Start(ctx context.Context) context.CancelFunc {
+	ctx, cancelFunc := context.WithCancel(ctx)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				// fallthrough
+			}
+
+			for _, path := range l.paths {
+				if !l.fileHasChanged(path) {
+					continue
+				}
+
+				err := l.sendLogEntries(path)
+				if err != nil {
+					logger.Warn("Failed sending log entry", logger.Ctx{"err": err})
+				}
+			}
+
+			time.Sleep(10 * time.Second)
+		}
+	}()
+
+	return cancelFunc
+}
+
+// fileHasChanged returns true if either the size or modtime of the given file has changed.
+func (l *OVNLogMonitor) fileHasChanged(filePath string) bool {
+	// Get the file stats. As a positive side effect, this will also return false if the file doesn't exist.
+	stat, err := os.Stat(filePath)
+	if err != nil {
+		return false
+	}
+
+	initialStat, ok := l.initialStats[filePath]
+	if !ok {
+		// If the stats haven't be recorded yet, mark the file as changed.
+		l.initialStats[filePath] = stat
+		return true
+	}
+
+	// If the stats have changed, mark the file as changed.
+	if stat.Size() != initialStat.Size() || stat.ModTime() != initialStat.ModTime() {
+		l.initialStats[filePath] = stat
+		return true
+	}
+
+	// File content is unchanged.
+	return false
+}
+
+// sendLogEntries sends the log entries to the events API. It only sends entries which haven't been sent already.
+func (l *OVNLogMonitor) sendLogEntries(filePath string) error {
+	content, err := os.ReadFile(filePath)
+	if err != nil {
+		return fmt.Errorf("Failed reading file %q", filePath)
+	}
+
+	scanner := bufio.NewScanner(bytes.NewReader(content))
+
+	for scanner.Scan() {
+		// Example log entry:
+		// 2023-08-30T05:23:38.425Z|00012|vlog|INFO|opened log file /var/log/ovn/ovn-controller.log
+		fields := strings.SplitN(scanner.Text(), "|", 5)
+
+		if len(fields) != 5 {
+			continue
+		}
+
+		timestamp, err := time.Parse(time.RFC3339Nano, fields[0])
+		if err != nil {
+			timestamp = time.Now()
+		}
+
+		// Ignore log entries which have already been sent, or are older than the last scan.
+		if timestamp.Before(l.lastTimestamp) {
+			continue
+		}
```

Comment:
Also I think that storing the last timestamp as a property of the OVNLogMonitor as a whole rather on a per-file basis could be problematic. I'm thinking of a scenario where 2 entries are logged at the same time in 2 different files. The first file is processed and records the last log time, and then moves onto the 2nd file and skips the entry because it has the same event time as the earlier processed log line.